aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -3708,4 +3723,96 @@ std::tuple<Tensor&,Tensor&> legacy_lstsq_out(
+Tensor _det_lu_based_helper_backward_helper(
+  const Tensor& det_grad,
+  const Tensor& det,
+  const Tensor& self,
+  const Tensor& lu,
+  const Tensor& pivs
+) {
+  auto eps = at::native::_get_epsilon(c10::toValueType(self.scalar_type()));
+  auto n = self.size(-1);
+  auto eps_tensor = at::tensor(eps, self.options());
+  auto condition_diagonal = [&](const Tensor& x) {
+    auto x_diag = x.diagonal(0, -2, -1);
+    auto x_diag_conditioned = at::where(
+      x_diag == 0.0,
+      eps_tensor,
+      x_diag
+    );
+    x_diag.copy_(x_diag_conditioned);
+  };
+
+  // create a matrix d := (det_grad * det.conj()) I
+  // NOTE: we do not use the shorter version
+  // auto d = at::zeros_like(self);
+  // d.diagonal(0, -2, -1).copy_((det_grad * det.conj()).unsqueeze(-1));
+  // to avoid in-place operations to eliminate potential issues with Vmap
+  auto det_expanded_sizes = det.sizes().vec();
+  det_expanded_sizes.push_back(n);
+  auto d_diag = det_grad * det.conj();
+  auto d = at::diag_embed(d_diag.unsqueeze(-1).expand(det_expanded_sizes));
+  // make sure that d is Fortran-contiguous. The transposition is sufficient as d is a diagonal square matrix
+  d = d.transpose(-2, -1);
+
+  if (self.device().type() == at::kCPU) {
+    // we want to condition the diagonal of the lu Tensor, but it is not allowed
+    // to modify arguments of backward functions in-place, hence the cloning.
+    auto lu_clone = lu.clone();
+    condition_diagonal(lu_clone);
+
+    auto trans = self.is_complex() ? 'C' : 'T';
+
+    // d is modified in-place and will contain the result
+    lu_solve_trans_stub(self.device().type(), d, lu_clone, pivs, trans);
+    return d;
+  }
+  // lu_solve is less stable than two triangular_solve for CUDA tensors.
+  else {
+    Tensor p, l, u;
+    std::tie(p, l, u) = at::lu_unpack(lu, pivs, /*unpack_data=*/true, /*unpack_pivots=*/true);
+
+    if (self.is_complex()) {
+      // Tensors u_h and l_h should be physically conjugated prior to applying kernel stubs,
+      // as .conj() is lazy and will not materialize conjugated output.
+      l.conj_physical_();
+      u.conj_physical_();
+    }
+
+    auto infos = at::zeros({std::max<int64_t>(1, batchCount(self))}, self.options().dtype(kInt));
+
+    // triangular_solve_stub performs operations in-place.
+    // Tensor d will contain the result
+    condition_diagonal(u);
+
+    // Solve u^h x = d
+    // note that d = c I for some scalar c, hence
+    // d u_h^{-1} = c I u_h^{-1} = u_h^{-1} c I = u_h^{-1} d.
+    // NOTE: u is contigious and upper-triangular,
+    // but from the Fortran respective it is lower-triangular and already transposed.
+    // Since u is conjugated in-place in the code above, it is sufficient
+    // to just run triangular_solve with upper=false.
+    triangular_solve_stub(
+      self.device().type(), u, d, infos,
+      /*upper=*/false,
+      /*transpose=*/false,
+      /*conjugate_transpose=*/false,
```

Comment:
How come we do not use this `= true` rather than using `conj_physical_()` above?
Same question for the next solve.

Reply:
Because in the comment it is explained why the result is Transposed in the eyes of Fortran, but we need to also conjugate it, but not transpose conjugate.

Reply:
Might it be better rather than doing `conj_physical_()` which incurs on a copy, to do `transpose(-2, -1)` and set `conjugate_transpose = true` then?
Even better (albeit a bit crazy) would be to see if we can simply omit the conjugate all together and do `transpose=true` AND `conjugate_transpose=true`. This should (at least semantically) be equivalent to saying that the input is conjugated, but it may not work. This API with `transpose` and `conjugate_transpose` as booleans is very weird really.

Reply:
`conj_physical_` is done in-place. The API is indeed confusing, as for real/complex tensors you have set both flags!

Reply:
I'm sorry, but can we make API less confusing? All we are doing with it is 
```
  magma_trans_t trans = transpose ? MagmaTrans : MagmaNoTrans;
  trans = conjugate_transpose ? MagmaConjTrans : trans;
```
so it would be less confusing to do it the way blas libraries do: have a char or enum with 3 values, NoTrans, Trans and ConjTrans. cc @IvanYashchuk